### DISCUSSION TITLE
Handler does not accept AST as a query

### DISCUSSION THREAD

**Discussion by ddiachkov** [enisdenjo/graphql-sse#65] [Q&A] [answered]
Hello!

I use graphql-codegen to parse and pre-compile queries into AST (`DocumentNode`). When I send the query to the server as JSON, I get `Invalid query` error due to [this check](https://github.com/enisdenjo/graphql-sse/blob/master/src/handler.ts#L1063).

It looks like both `parseReq` and `prepare` are designed[^1][^2] to handle this properly. So I think this is a bug.

**Expected behavior**: no error; handler uses provided AST
**Current behavior**: `{"errors":[{"message":"Invalid query"}]}`

Example cURL:

```
curl -v -X POST 'http://0.0.0.0:8000/graphql/kiosk/stream' -H 'accept: text/event-stream' -H 'Accept-Encoding: gzip' -H 'authorization: xxx' -H 'Connection: Keep-Alive' -H 'Content-Length: 505' -H 'Content-Type: application/json; charset=utf-8' -H 'Host: 0.0.0.0:8000' -H 'User-Agent: okhttp/5.0.0-alpha.11' -H 'x-snackpass-client: @snackpass/kiosk' -H 'x-snackpass-client-version: 32.5.0' -d '{"query":{"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"subscription","name":{"kind":"Name","value":"SubscribeToActiveOrders"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"activePurchases"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"_id"}},{"kind":"Field","name":{"kind":"Name","value":"status"}},{"kind":"Field","name":{"kind":"Name","value":"timestamp"}}]}}]}}]}}'
```

[^1]: https://github.com/enisdenjo/graphql-sse/blob/master/src/handler.ts#L1038
[^2]: https://github.com/enisdenjo/graphql-sse/blob/master/src/handler.ts#L610

**Comment by enisdenjo** [comment, marked as answer]
It's not a bug, the [GraphQL over HTTP spec requires the `query` field to be a string](https://graphql.github.io/graphql-over-http/draft/#sel-FALFFBDAAACFBtjd).

You can either:
1. Use the [`string` documentMode for the client preset in GraphQL Codegen](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#documentmode)
2. [`print` the AST](https://graphql.org/graphql-js/language/#print) before sending it

**Reply by ddiachkov** [reply]
Yes, you're absolutely right!

The fact that `RequestParams.query` is defined as `DocumentNode | string` misled me.

**Reply by enisdenjo** [reply]
Oh, that was definitely a wrong type definition. I've fixed it in https://github.com/enisdenjo/graphql-sse/commit/16c9600b37ca21d6cfa3fa745bd41887b98589fd. Thanks for pointing it out!
